### ISSUE TITLE
About Sections of content

### ISSUE THREAD

**Issue by Jym77** [act-rules/act-rules.github.io#1311]
There's been quite a lot of discussion going around this topic, and I feel we might need more… Trying to gather here what have been said so far, as well as some extra personal thoughts.

Discussion was in #1159, #1293, and 14/05/20 CG call (minutes not there yet).
Tagging @carlosapaduarte, @daniel-montalvo, @EmmaJP, @kasperisager, @ShadowBB, @WilcoFiers who took part in the discussion at some point, but any input is obviously welcome.
We might want to involve people from the AGWG and get input from them.

As far as we are concerned, the problem is currently restricted to rule [Heading is descriptive](https://act-rules.github.io/rules/b49b2e) which is the only one using sections of content, but it can be more generic (notably, #421 is also needing sections of content). So we shouldn't focus too much on this specific rule, even though we shouldn't over-engineer the definitions for uses we don't have yet…

---
## Are there different types of Sections of content? Of headings?
WCAG is using the term "heading" in a generic sense, without making any distinction between, say, "elements with a role of `heading`" and "elements styled with a larger font". Should we make such a distinction?

I feel there is no problem in us splitting the WCAG concept of "heading" into sub-concepts if it helps write rules (mostly by keeping rules focused). Of course, we don't want to artificially create too many rules.
If we feel that a "heading" can be either "heading of type A" or "heading of type B" (maybe sometimes both), I think there is no problem in writing a rule "heading of type A is descriptive" given that yes, WCAG do require "headings of type A" to be descriptive (because "headings of type A" are "headings" and thus need to be descriptive).

This is a bit similar to what we do for 1.1.1. WCAG requires all non-text content to have an alternative, and does not really have different type of non-text content (as far as 1.1.1 is concerned). Yet, we write rules for images and videos separately (and even separate `img` HTML images, and SVG images…) This is because an image is non-text content and thus require an alternative, so a rule requiring "image has an accessible name" is testing a WCAG SC.

So, if splitting the concept of "heading" into subconcepts can help us writing rules, let's do it. We may end up with partial checks only (especially if some of the subconcepts can't be defined in a satisfactory way), but we'll nonetheless end up with good rules.

---
## About covariance and contravariance
There is a big caveat with the use of subconcepts when writing rules: we can only use them in a covariant context, never in a contravariant one. This is to preserve the fail-on-fail relationship.

Consider "image" as a subconcept of "non-text content". A rule "image has accessible name" uses this in a **covariant** way. Thus, the set of errors flagged by this rule is a subset of all errors of 1.1.1 and this is good.
Now, let's suppose that there was a SC "only non-text content has an accessible name". Writing a rule "element with accessible name is an image" would use "image" as a subconcept of "non-text content" in a **contravariant** way. This result in the set of errors flagged by this rule being a **superset** of the actual (well, hypothetical) errors, thus breaking the fail-on-fail relationship.
Compare with the use of "non-empty accessible name" (in "image has non-empty accessible name"). This is a super-concept of "text alternative that serves the equivalent purpose". But it is used (in this rule) in a contravariant way. Using a super-concept in a contravariant way results in the set of detected errors being a subset of the actual errors, which is why we want.

**Subconcepts can only be used covariantly, super-concepts can only be used contravariantly.**

Going back to heading and sections. One way to define sections is "between a start heading and a end heading". In this definition, "start heading" is used covariantly (if we only consider some kind of headings, not all, we will detect only a subset of all sections). But "end heading" is used contravariantly (if we only consider some kind of headings, the detected sections will be larger than the actual ones).
This is illustrated by the following examples (considering "element with a role of `heading`" as a subset of "heading"):
```html
<div class="styled-as-h1">Lemon</div>
This fruit is yellow.
<h1>Lime</h1>
This fruit is green.
```
Here, we would not detect the "Lemon" section but that wouldn't be a problem. What is detected as a section is indeed a section and if we detect a non-descriptive heading, this is indeed a failure of 2.4.6

```html
<h1>Lime</h1>
This fruit is green.
<div class="styled-as-h1">Lemon</div>
This fruit is yellow.
```
Here, however, we don't detect the end of the "Lime" section. This results in putting into the "Lime" section too much content and thus we'll flag an error when there is (arguably) none (for 2.4.6). The contravariant "end heading" cannot be restricted to a subconcept.

---
## Which SC is failing?
An additional difficulty comes from the fact that 1.3.1 and 2.4.6 are separated. 2.4.6 concerns essentially "stuff that is perceived as heading" (as far as we're concerned so far). So we cannot rely on correct markup (i.e. role of `heading`) to detect headings. "Element with a role of `heading`" is a subconcept of "heading" (even though it might be styled in a way that sighted users won't perceived it as a heading, it will still be perceived as a heading by users of AT), and if we can build a rule "element with role of heading is descriptive" because this uses the subconcept in a covariant way.

Going back to the previous example
```html
<h1>Lime</h1>
This fruit is green.
<div class="styled-as-h1">Lemon</div>
This fruit is yellow.
```
We could consider that there is only one section of content, with heading "Lime", and that this heading is not descriptive, thus failing 2.4.6. This is probably what is going to be perceived by users of AT. However, the (likely) fix for this case is to give a role of heading to the `div` (or to use an actual `h1`…) I.e. the fix is a 1.3.1 fix (have semantic match visual design) and not a 2.4.6 fix (update heading).

On the other hand,
```html
<div role="heading">Lime</div>
This fruit is red.
<div class="styled-as-h1">Lemon</div>
This fruit is blue.
```
is failing both 1.3.1 (for each `div`), and 2.4.6 (for each heading). The fix is not just a 1.3.1 fix (having design and semantics match), but also a 2.4.6 fix (having heading and content match).

This adds difficulty to the rule and definition. If we assume that each "heading" has a role of `heading` (hence that 1.3.1 is satisfied as far as heading are concerned, a consequent but not irrealistic assumption), then "element with a role of `heading`" become synonym with "heading" and can be used in co- and contravariant contexts! This might be a way out of the full problem…

---
## Visual design vs semantics
One of the important discussion points… Does the structure of the page stem from the visual design or from the programmatic structure? It is clear that if they don't match, then the page fails 1.3.1. But many SC (including 2.4.6 which concerns us here) are independant of 1.3.1. It is possible to fail 1.3.1 and pass 2.4.6, it is possible to fail 2.4.6 and pass 1.3.1.

On one hand, it is clear that we can build any design with only `div`s and `span`s (and that most of the web was done that way back in the days…) Using larger font, colors, borders, … a rich structure can be conveyed this way.
Even further, sighted users only perceive the visual design. I do not perceive a `h1` as a "heading" because it has a role of `heading`, but because it is visually presented to make me think it's a heading (using standard code). It is somewhat "lucky" that the UA chose to render `h1` elements in a way that make me perceived them as "headings" (of course, it's no luck but deep designs, whatever…)
Even when the semantics seems to have a visible (as in, literally visible) effect, it is still not what I perceive. If I click on a `label` element, focus will move to the associated form field. This looks like a visible effect of the semantics of `label` element, but it's just a layer put on top of it (like rendering `h1` with bigger font is just a layer put on top of the semantics). The same effect could be achieved with a `<span onlcick="target.focus()">` and I would never know whether the semantics association has been properly done or not.

So, in that sense, the visual design is all that matter.

On the other hand, users of AT (well, of screen readers at least) don't care about visual design. Technically, they do not either have direct access to the (programmatic) semantic, only to what the AT is exposing to them. An AT could very well have a heuristic for determining "headings" (e.g. by font size) and list these in its "list of headings" (instead of listing elements with a role of `heading`). 

Yet, the programmatic semantic is meant to be used. By AT, of course. But also by visual UAs to create a visual design that matches it. Essentially, if a dev uses semantic HTML and does't override style, the default UA style and actions will turn the semantic into acceptable visual design, and 1.3.1 should pass.

Lastly, the rule we're looking at has always been targeting "elements with a role of `heading` that are visible **or** in the accessibility tree" thus indicating that the rule has always considered semantic as important (maybe as a way to focus on an unambiguous subset of all headings only).

Thus, I am reluctant to consider the visual design only. This feels like considering that sighted users are first class users and that users of AT come after.

Additionally, looking at the previous example,
```html
<h1>Lime</h1>
This fruit is green.
<div class="styled-as-h1">Lemon</div>
This fruit is yellow.
```
for sighted users, this is not failing 2.4.6, but for screen readers users, this is. The failure of 2.4.6 for screen readers users stem from the failing of 1.3.1, which is why we (maybe) consider that this only fails 1.3.1 but not 2.4.6

But we can make the symmetrical example:
```html
<h1>Lime</h1>
This fruit is green.
<div role="heading" aria-level=1>Lemon</div>
This fruit is yellow.
```
Now, screen readers users will see a perfectly well structured document that doesn't fail 2.4.6. But sighted users will only perceived one heading and section (arguably, let's say it's the case for the example…) and thus a non-descriptive heading. Again, the fix here is (likely) to style the `div` correctly (or use an actual `h1`…), i.e. a 1.3.1 fix, not a 2.4.6 fix.

So, I feel that basing everything on the visual design only is not more correct than basing everything on semantics only. Yes, "visual headings" have to be descriptive, but they form a subset of all headings and we must take care of using these only in a covariant context. Notably, we can't have sections of content "from a start visual heading to a end visual heading" as the second use is contravariant.

---
## What is section of content
Of course, one of the key point is here…
My main problem is that I believe that "is X a section of content" is  indecidable in general (I do have a strong computability background…) To put it otherwise, this probably means that the notion is highly subjective. We can definitely agree that some precise examples are sections of content, but it will be hard to get an exact definition we'll all agree with.
Furthermore, "section of content" is not only subjective on the user part, but the technologies themselves add a layer of "subjectivity". As we've seen with the examples above, what sighted users and screen readers users perceived as "section of content" might be different. Essentially, 1.3.1 is precisely here to patch that "technological subjectivity". But again, 2.4.6 can pass while 1.3.1 is failing, so we have to take that subjectivity into account.
Even worse, the UA could have a bad default rendering. For example, not making the heading any different from the rest of the text, thus destroying any "heading to heading" visual section. That is also part of the technological subjectivity we should keep in mind (and maybe discard if it's too corner case).

And again, coming from computability, the way to tackle undecidable sets is sometimes to go with sub/super sets (approximations). We might be able to come up with a (objective) definition of "section of content of type A" which we agree is a subset of everything we (personally) consider as section of content. These could be used in a covariant context: anything that WCAG requires of "section of content", it also requires of "sections of content of type A".
Similarly, we can come up with an objective super-set of sections of content. Something every body agree that "yeah, any section of content will verify these properties". This super-set could be used in a contravariant context.

Unfortunately, it is easy to mistreat HTML badly. This makes both sub- and super-sets of sections of content hard to define… We've seen that "heading to heading" is only as good as the definition of heading (and is actually not correct because the last section of content might stop at the `footer` element).

Even landmarks or `section` elements are not good. Sure, they'll likely be perceived as sections of content by AT users, but not necessarily by sighted users:
```html
<h1>Lemon</h1>
<nav>This fruit is yellow</nav>
<p>This fruit is acid</p>
```
The `nav` is not going to be perceived as a section of content by sighted users. So `nav` elements can't really be a subset of "section of content". OTOH, we can consider that the `nav` is going to be perceived as a section of content by some users (those who can perceived or navigate landmarks), so `nav` elements are a subset of "section of content". 
There is a related question about subconcept: In order to be a section of content, does something need to be perceived as one by **all** users, or by **some** (specific, e.g. sighted) users, or by **any** (non-specific) user?
I feel that stuff does not need to be perceived as section of content by **all** users in order to be one, because 1.3.1 is exactly here to say that all users have the same perception (so if "**all** users" was needed, 1.3.1 would always be satisfied).
I don't know about the rest. Maybe the question here would be:
Is a `nav` element **always** a section of content by itself?

It's not even clear that sections of content group elements that are adjacents (in the DOM tree). I can abuse a bunch of `div` with absolute positioning so that the rendered structure is very different from the DOM structure. This could result in visual sections of content that are not contiguous elements in the DOM tree. This illustrate that a super-set of section of content is not easy to build either.

---
## What is heading
Of course, sections of content are a side-artefact for 2.4.6. 2.4.6 is not about sections of content, it is about headings. They must "describe topic or purpose". We kinda need section of content because WCAG doesn't tell **of what** must a heading describe topic or purpose.

The same issues come for headings. I think it is also a highly subjective (undecidable) notion. Fortunately, it is a bit easier to have some decidable (objective) subsets of headings. Most notably, "the set of elements with a role of `heading`" is a subset of all headings. Maybe.

This comes back to the previous question. Elements with a role of `heading` will be perceived as headings by some users. Not by all. Is it enough to make them "headings"?
Again, we have a similar question: is an element with a role of `heading` **always** a "heading" (in the WCAG sense of the term)? If yes, then we have a decidable subset. If no, …

And again, a subset of headings can only be used in a covariant context. So "from heading to heading" is not good and must be "from (subset of heading) to (superset of heading)".

Ultimately, of course, "headings" are just "what users perceive as heading" which is a very subjective definition but could work. The problems with such a definition are:
* It doesn't bring any disambiguation from WCAG. A rule saying that "stuff perceived as heading describe topic or purpose" doesn't bring any value on top of WCAG and thus is kinda useless for us (we're not here just to repeat WCAG…) [and the "of what?" question remains very ambiguous in any case]
* Being highly subjective, it can't be used in Applicability. Of course, we could circumvent that by having a rule targeting the full document and requiring that every "stuff perceived as heading" is descriptive, thus moving the subjective definition to the expectation. But this is annoying for error reporting (the rule would say "one of your heading is bad, good luck finding which) while a rule targeting the heading is more precise (and says "this precise heading here is bad, fix it").

---
## What to do?
Of course, this is the hardest question and I'm absolutely not sure of the answer.

I'm not convinced by the "pure visual" approach which I tried a bit in #1159, for three reasons. 
First, as the discussion there shows, it is very easy to forget some case in the definition of which visual delimiters to consider and this result in some case being incorrectly ignored. We might be able to build something that is a subset of what we approximate, and to make sure that we only use the subconcept in a covariant way, but that requires a lot of work and is prone to errors.
Second, I feel this will stay subjective, hence not for applicability, hence not good for headings as explained above. And mixing a programmatic subset of headings with a visual subset of sections of contents seems very prone to errors, too…
Lastly, basing everything on the visual design feels like having first class users that can perceive this, and the rest being dependant on 1.3.1 passing to convey them the visual design. I don't like the idea of having first class users in our rules.

I'm also very reluctant in swapping everything to visual design-based definition because the current rule is absolutely not about this (applicability: "element with a role of `heading` that is visible or in the accessibility tree"). So this would be writing a full new rule (and deprecating the existing one) rather than patching an existing one.

For me, the key question (of easily solving this puzzle) comes back to these examples:
```html
<h1>Lime</h1>
This fruit is green.
<div class="styled-as-h1">Lemon</div>
This fruit is yellow.
```
```html
<h1>Lime</h1>
This fruit is green.
<div role="heading" aria-level=1>Lemon</div>
This fruit is yellow.
```
Are these only failing 1.3.1 or also 2.4.6? For some users, there are two sections of content here, for 
some there is only one (section of content is highly subjective…) To stay in terms of 2.4.6, for some users there are two headings and they "describe topic" (of what?) but for some users, there is only one heading and it does not "describe topic".

If we consider that the examples above **also** fail 2.4.6, i.e. that 2.4.6 means "if any user perceives something as a section of content, and something as its heading, then the thing perceived as heading must describe the thing perceived as section of content", then there is a relatively easy way out of it:
* consider that section of content goes from "element with a role of `heading` or start of a landmark" to "element with a role of `heading` or start of a landmark". The reason for this being that both elements with a role of `heading` and start of landmarks are announced by AT (at least by NVDA with default settings on Chrome). Thus, users of these AT will perceive these, and perceive them in a way that indicate a change of section. So, stuff between these will be perceived as a full section by some users, and if there is a heading for it, it must be descriptive.

Of course, this is going to flag the examples above. This can be a bit alleviated by adding an assumption that stuff perceived as heading (by any user) is correctly marked up with a role of `heading` (i.e. that 1.3.1 is satisfied as far as headings are concerned). In which case the example above with `<div class="styled-as-h1">` would be ruled out as not matching the assumption.

This still leaves possible gap with things like:
```html
<div class="card">
  <h1>Lime</h1>
  This fruit is green.
</div>
<div class="card">
  Lemons are yellow.
</div>
```
where the `card` class can be styled to make it obvious for sighted users that it is a completely different topic.
Arguably, this still fails 2.4.6 because users of AT will perceive only one section with content "Lime heading level 1 This fruit is green. Lemon are yellow." and the heading does not describes what is perceived as a single content…
But maybe that only fails 2.4.10…
Maybe here also this can be alleviated by assuming that 2.4.10 is already satisfied (i.e. that "content has been organised into sections using headings"). Then this latest example would be ruled out by the assumption.

---
## TLDR: An easy but maybe too restrictive proposition
So, based on what I say just above, an easy way to patch the rule would be:
* Assume that "headings have been used to organise content into sections" (i.e. that the page passes 2.4.10).
* Assume that "headings have a role of `heading`" (i.e. that the page passes 1.3.1 as far as headings are concerned
* Check whether "elements with role of `heading`" describe "content until next element with role of `heading` (and lower or equal level)". Given the two assumptions above, this would indeed detect failures of 2.4.6. This essentially consider that a section of content stops when the next one starts.

Question is whether such assumptions would be too broad (and excluding too many pages from the rule)… Notably the 2.4.10 assumption would require `footer`s to have heading… Can be tweaked by checking instead "elements with role of `heading`" describe "content until next element with role of `heading` (and lower or equal level) or `landmark`" but that's not super clean…

Thoughts?

**Comment by Jym77** [comment]
## A Litmus test question
Maybe a question that might help us find a way…
Consider these two examples:
```html
<h1>Lime</h1>
This fruit is green.
<div class="styled-as-h1">Lemon</div>
This fruit is yellow.
```
```html
<h1>Lime</h1>
This fruit is green.
<div role="heading" aria-level=1>Lemon</div>
This fruit is yellow.
```
They are both failing 1.3.1.
Are they failing 2.4.6? If your answer is different for both examples, why?

**Comment by kasperisager** [comment]
Regarding those two cases, I don't think it's controversial that both cases pass 2.4.6. From the WCAG, emphasis mine:

> This Success Criterion does not require headings or labels. This Success Criterion requires that if headings or labels are provided, they be descriptive. **This Success Criterion also does not require that content acting as a heading or label be correctly marked up or identified - this aspect is covered separately by 1.3.1: Info and Relationships.** It is possible for content to pass this Success Criterion (providing descriptive content that acts as headings or labels) while failing Success Criterion 1.3.1 (if the headings or labels aren't correctly marked up/identified). Conversely, it is also possible for content to pass Success Criterion 1.3.1 (with headings or labels correctly marked up or identified), while failing this Success Criterion (if those headings or labels are not sufficiently clear or descriptive).

https://www.w3.org/WAI/WCAG21/Understanding/headings-and-labels.html

**Comment by kasperisager** [comment]
> So, based on what I say just above, an easy way to patch the rule would be:
> 
> * Assume that "headings have been used to organise content into sections" (i.e. that the page passes 2.4.10).
> * Assume that "headings have a role of `heading`" (i.e. that the page passes 1.3.1 as far as headings are concerned
> * Check whether "elements with role of `heading`" describe "content until next element with role of `heading` (and lower or equal level)". Given the two assumptions above, this would indeed detect failures of 2.4.6. This essentially consider that a section of content stops when the next one starts.

This is exactly what I had in mind myself while reading your notes so +1 on that. Kudos by the way, it's a great writeup! 👏

**Comment by Jym77** [comment]
> ## TLDR: An easy but maybe too restrictive proposition
> So, based on what I say just above, an easy way to patch the rule would be:
> 
> * Assume that "headings have been used to organise content into sections" (i.e. that the page passes 2.4.10).
> * Assume that "headings have a role of `heading`" (i.e. that the page passes 1.3.1 as far as headings are concerned
> * Check whether "elements with role of `heading`" describe "content until next element with role of `heading` (and lower or equal level)". Given the two assumptions above, this would indeed detect failures of 2.4.6. This essentially consider that a section of content stops when the next one starts.

Of course, a problem with that (or any "heading-to-something" section of content) is that it is not usable for a 2.4.10 rule. Essentially, a definition assuming that 2.4.10 passes cannot be used to check 2.4.10…
This is OK for the "heading is descriptive" rule (or other 2.4.6 rule). After all, 2.4.6 is only applicable if headings are used in the first place…

On a larger scale, not being able to write a 2.4.10 rule is mildly annoying (it is only a AAA SC).
But this also makes the definition unusable for 2.4.1 (Bypass Blocks) since having headings at the start of sections is a way to pass it…
But I think it is still OK to keep that definition for "heading is descriptive" and work out something else for Bypass blocks. At least because "heading is descriptive" needs a heading-section association which is immediate with this definition but more complex with another, not based on headings.

**Comment by Jym77** [comment]
I think there are 2 different questions here.

One is the full-fledged "what is a section of content?" and the other is "how to fix rule Heading is descriptive?" Given the applicability of the rule, the answers might be very different. I've split the second question to #1361 which is probably the problem we need to tackle.

**Comment by Jym77** [comment]
The more I'm thinking about "sections of content" in general, the more I think they are defined by, well, content…

A long text with many paragraphs talking about different topics will likely be perceived as having several sections of content and lacking headings, visual separation, … It might not be 100% clear where the sections of content stop and start, but it will likely still be clear that there are several sections about different topics.

Not sure if that helps…

**Comment by ajanec01** [comment]
If the definition is supposed to fit only the elements with the role of heading then a section of content could be changed to something like “programmatically determined section context”.  I base that on the definition of [programmatically determined link context](https://www.w3.org/TR/WCAG21/#dfn-programmatically-determined-link-context). I see an analogy between the two definitions. Link in context also referred to a certain section of a page
